Source/Rampage/ViewController.swift

```diff
@@ -26,68 +26,128 @@ private func loadTextures() -> Textures {
+    }
+    
     override func viewDidLoad() {
         super.viewDidLoad()
-        setUpImageView()
-
+        
         let displayLink = CADisplayLink(target: self, selector: #selector(update))
         displayLink.add(to: .main, forMode: .common)
-
-        view.addGestureRecognizer(panGesture)
+        
     }
-
-    private var inputVector: Vector {
-        switch panGesture.state {
-        case .began, .changed:
-            let translation = panGesture.translation(in: view)
-            var vector = Vector(x: Double(translation.x), y: Double(translation.y))
-            vector /= max(joystickRadius, vector.length)
-            panGesture.setTranslation(CGPoint(
-                x: vector.x * joystickRadius,
-                y: vector.y * joystickRadius
-            ), in: view)
-            return vector
-        default:
-            return Vector(x: 0, y: 0)
-        }
-    }
-
+    
     @objc func update(_ displayLink: CADisplayLink) {
         let timeStep = min(maximumTimeStep, displayLink.timestamp - lastFrameTime)
-        let inputVector = self.inputVector
-        let rotation = inputVector.x * world.player.turningSpeed * worldTimeStep
+        let leftInputVector = self.contentView.leftJoystickInputVector
+        let rightInputVector = self.contentView.rightJoystickInputVector
+        
+        let rotation = rightInputVector.x * world.player.turningSpeed * worldTimeStep
         let input = Input(
-            speed: -inputVector.y,
+            speed: leftInputVector.orthogonal,
             rotation: Rotation(sine: sin(rotation), cosine: cos(rotation))
         )
         let worldSteps = (timeStep / worldTimeStep).rounded(.up)
         for _ in 0 ..< Int(worldSteps) {
             world.update(timeStep: timeStep / worldSteps, input: input)
         }
         lastFrameTime = displayLink.timestamp
-
+        
         let width = Int(imageView.bounds.width), height = Int(imageView.bounds.height)
         var renderer = Renderer(width: width, height: height, textures: textures)
         renderer.draw(world)
-
+        
         imageView.image = UIImage(bitmap: renderer.bitmap)
     }
+    
+}
 
-    func setUpImageView() {
-        view.addSubview(imageView)
-        imageView.translatesAutoresizingMaskIntoConstraints = false
-        imageView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
-        imageView.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
-        imageView.widthAnchor.constraint(equalTo: view.widthAnchor).isActive = true
-        imageView.heightAnchor.constraint(equalTo: view.heightAnchor).isActive = true
+class ContentView: UIView {
+    
+    private(set) lazy var imageView: UIImageView = {
+        let imageView = UIImageView()
         imageView.contentMode = .scaleAspectFit
         imageView.backgroundColor = .black
         imageView.layer.magnificationFilter = .nearest
+        return imageView
+    }()
+    
+    private let leftJoyStick = UIView()
```

Comment:
I would suggest you create a new `UIView` subclass called `JoystickView` (or whatever) and move all of the gesture logic into that, so that you only need to create two view instances in the view controller instead of two copies of everything.